src/config/mod.rs

```diff
@@ -287,54 +308,70 @@ pub enum LogTarget {
 //----------- Setting ----------------------------------------------------------
 
 /// A configured setting.
-#[derive(Debug, Clone, Copy)]
+#[derive(Debug, Default, Clone, Copy)]
 pub struct Setting<T> {
-    /// The source of the value.
-    pub source: SettingSource,
+    /// The default for the value.
+    pub default: T,
 
-    /// The underlying value.
-    pub value: T,
+    /// The setting in the configuration file, if any.
+    pub file: Option<T>,
+
+    /// The setting from environment variables, if any.
+    pub env: Option<T>,
+
+    /// The setting from command-line arguments, if any.
+    pub args: Option<T>,
 }
 
 impl<T> Setting<T> {
-    /// Merge two [`Setting`]s, keeping the highest-priority value.
-    pub fn merge(&mut self, other: Self) {
-        if self.source < other.source {
-            self.value = other.value;
+    /// Construct a new [`Setting`].
+    pub const fn new(default: T) -> Self {
+        Self {
+            default,
+            file: None,
+            env: None,
+            args: None,
         }
     }
 
-    /// Merge a [`Setting`] with a value from a fixed source.
-    pub fn merge_value(&mut self, value: Option<T>, source: SettingSource) {
-        if let Some(value) = value {
-            self.merge(Setting { source, value });
+    /// The current value.
+    pub const fn value(&self) -> &T {
+        match self {
+            Self {
+                args: Some(value), ..
+            }
+            | Self {
+                env: Some(value), ..
+            }
+            | Self {
+                file: Some(value), ..
+            }
+            | Self { default: value, .. } => value,
```

Comment:
I wonder if
```rust
self.args
    .or(self.env)
    .or(self.file)
    .unwrap_or(self.default)
```
is slightly more explicit about the ordering (even though it's functionally the same).

Reply:
I love this system btw, to be able to see all of these values and tell the user where the things come from.